src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.FullGlobalizationData.cs

```diff
@@ -7,260 +7,75 @@ namespace System
-
-            // Some edge cases only apply when the offset is +00:00.
-            if (baseUtcOffset == TimeSpan.Zero)
+            if (windowsId.Equals("utc", StringComparison.OrdinalIgnoreCase))
             {
-                // GMT and its aliases will just use the equivalent of "Greenwich Mean Time".
-                string? gmtLocationName = null;
-                GetDisplayName(GmtId, Interop.Globalization.TimeZoneDisplayNameType.GenericLocation, uiCulture.Name, ref gmtLocationName);
-                if (genericLocationName == gmtLocationName)
-                {
-                    displayName = $"{baseOffsetText} {genericName}";
-                    return;
-                }
-
-                // Other zones with a zero offset and the equivalent of "Greenwich Mean Time" should only use the location name.
-                // For example, prefer "Iceland Time" over "Greenwich Mean Time (Reykjavik)".
-                string? gmtGenericName = null;
-                GetDisplayName(GmtId, Interop.Globalization.TimeZoneDisplayNameType.Generic, uiCulture.Name, ref gmtGenericName);
-                if (genericName == gmtGenericName)
-                {
-                    displayName = $"{baseOffsetText} {genericLocationName}";
-                    return;
-                }
+                // Special case UTC, as previously ICU would convert it to "Etc/GMT" which is incorrect name for UTC.
+                ianaId = "Etc/UTC";
+                return true;
             }
 
-            if (genericLocationName == genericName)
+            foreach (char c in windowsId)
             {
-                // When the location name is the same as the generic name,
-                // then it is generally good enough to show by itself.
-
-                // *** Example (en-US) ***
-                // id                   = "America/Havana"
-                // baseOffsetText       = "(UTC-05:00)"
-                // standardName         = "Cuba Standard Time"
-                // genericName          = "Cuba Time"
-                // genericLocationName  = "Cuba Time"
-                // exemplarCityName     = "Havana"
-                // displayName          = "(UTC-05:00) Cuba Time"
-
-                displayName = $"{baseOffsetText} {genericLocationName}";
-                return;
-            }
-
-            // Prefer location names in some special cases.
-            if (StringArrayContains(timeZoneId, s_ZonesThatUseLocationName, StringComparison.OrdinalIgnoreCase))
-            {
-                displayName = $"{baseOffsetText} {genericLocationName}";
-                return;
+                // ICU uses some characters as a separator and trim the id at that character.
+                // while we should fail if the Id contained one of these characters.
+                if (c == '\\' || c == '\n' || c == '\r')
```

Comment:
I see this for validation of the chars duplicated can we make it a helper Private method with aggressive inlining or would that cause extra allocations?

Reply:
it is not worth it. we can factor that if we need to add more.